bubuku/features/rebalance/change.py

```diff
@@ -0,0 +1,248 @@
+                continue
+            targets = self._list_active_brokers_with_skip(broker.broker_id)
+            if not force:
+                targets = [t for t in targets if t.has_free_replica_slots()]
+            for _ in range(0, to_move):
+                copied = False
+                for topic_partition in broker.list_replicas():
+                    if self._move_partition_to_replica(broker, topic_partition, targets):
+                        copied = True
+                        break
+                if not copied:
+                    return False
+                if not force:
+                    targets = [t for t in targets if t.has_free_replica_slots()]
+        return True
+
+    @staticmethod
+    def _build_leader_distribution_weights(brokers: dict):
+        candidates = {}
+        for source_broker_id, source_broker in brokers.items():
+            if not source_broker.have_extra_leaders():
+                continue
+            for target_broker_id, target_broker in brokers.items():
+                if not target_broker.have_less_leaders():
+                    continue
+                weight_list = []
+                for topic in source_broker.topic_cardinality.keys():
+                    delta = source_broker.topic_cardinality[topic] - target_broker.topic_cardinality.get(topic, 0)
+                    weight_list.append((topic, delta))
+                # Now the first element is the first to rebalance.
+                if weight_list:
+                    candidates[(source_broker_id, target_broker_id)] = sorted(
+                        weight_list, key=lambda x: x[1], reverse=True)
+        return candidates
+
+    @staticmethod
+    def _place_to_cardinalities(array: list, topic: str, cardinality_change: int):
+        old_cardinality = None
+        for item in [(a, b) for a, b in array if a == topic]:
+            old_cardinality = item[1]
+            array.remove(item)
+        if old_cardinality is None:
+            return
+        new_cardinality = old_cardinality + cardinality_change
+        idx = len(array)
+        for i in range(idx - 1, -1, -1):
+            if array[i][1] >= new_cardinality:
+                idx = i + 1
+                break
+        array.insert(idx, (topic, new_cardinality))
+
+    @staticmethod
+    def rebalance_leaders(brokers: dict) -> list:
+        candidates = OptimizedRebalanceChange._build_leader_distribution_weights(brokers)
+        queue_list = []
+        while any([broker.have_extra_leaders() for broker in brokers.values()]):
+            top_candidate = None
+            for broker_pair, weight_list in candidates.items():
+                if not top_candidate:
+                    top_candidate = broker_pair
+                else:
+                    if weight_list[0][1] > candidates[top_candidate][0][1]:
+                        top_candidate = broker_pair
+            # Taking best topic to move
+            topic, _ = candidates[top_candidate][0]
```

Comment:
For example, up to here what it did was to get the best topic to move. I think it would be easier to verify correctness if there was a function for that which we exhaustively tested. The code up to here is simple, but when looking at it as whole, it's complicated, even with the comments to give some explanation and space to breath.